File: src/fable-library-py/fable/long.py
from typing import Optional


def fromBits(lowBits: int, highBits: int, unsigned: bool):
    ret = lowBits + (highBits << 32)
    if ret > 0x7FFFFFFFFFFFFFFF:
        return ret - 0x10000000000000000

    return ret


def fromInt(x):
    return x


def op_LeftShift(self, numBits):
    return self << numBits


def op_Addition(a, b):
    return a + b


def op_Multiply(a, b):
    return a * b


def op_UnaryNegation(value):
    return -value


def parse(string: str, style: int, unsigned: bool, _bitsize: int, radix: Optional[int] = None):
    return int(string)
    # res = isValid(str, style, radix)
    # if res:
    #     def lessOrEqual(x: str, y: str):
    #         length = max(len(x), len(y))
    #         return x.padStart(len, "0") <= y.padStart(len, "0");

    #     isNegative = res.sign == "-"
    #     maxValue = getMaxValue(unsigned or res.radix != 10, res.radix, isNegative);
    #     if (lessOrEqual(res.digits.upper(), maxValue)):
    #         string = res.sign + res.digits if isNegative else res.digits
    #         return LongLib.fromString(str, unsigned, res.radix);

    # raise Exception("Input string was not in a correct format.");


def toString(x):
    return str(x)

File: src/fable-compiler-js/fable-library/Long.js
import { getPrefix, isValid } from "./Int32.js";
import { fromString } from "./BigInt.js";
function getMaxValue(unsigned, radix, isNegative) {
    switch (radix) {
        case 2: return unsigned ?
            "1111111111111111111111111111111111111111111111111111111111111111" :
            (isNegative ? "1000000000000000000000000000000000000000000000000000000000000000"
                : "111111111111111111111111111111111111111111111111111111111111111");
        case 8: return unsigned ?
            "1777777777777777777777" :
            (isNegative ? "1000000000000000000000" : "777777777777777777777");
        case 10: return unsigned ?
            "18446744073709551615" :
            (isNegative ? "9223372036854775808" : "9223372036854775807");
        case 16: return unsigned ?
            "FFFFFFFFFFFFFFFF" :
            (isNegative ? "8000000000000000" : "7FFFFFFFFFFFFFFF");
        default: throw new Error("Invalid radix.");
    }
}
export function parse(str, style, unsigned, _bitsize, radix) {
    const res = isValid(str, style, radix);
    if (res != null) {
        const lessOrEqual = (x, y) => {
            const len = Math.max(x.length, y.length);
            return x.padStart(len, "0") <= y.padStart(len, "0");
        };
        const isNegative = res.sign === "-";
        const maxValue = getMaxValue(unsigned || res.radix !== 10, res.radix, isNegative);
        if (lessOrEqual(res.digits.toUpperCase(), maxValue)) {
            str = getPrefix(res.radix) + res.digits;
            str = isNegative ? res.sign + str : str;
            return fromString(str);
        }
    }
    throw new Error(`The input string ${str} was not in a correct format.`);
}
export function tryParse(str, style, unsigned, bitsize, defValue) {
    try {
        defValue.contents = parse(str, style, unsigned, bitsize);
        return true;
    }
    catch {
        return false;
    }
}
